Driver-only hashes: RSA 1.5 and PK + strategy doc (#6065)
Mbed-TLS/mbedtls: docs/architecture/psa-migration/strategy.md

`mbedtls_md` directly.
- Step 2 is achieved for most of X.509 and TLS (same gaps as step 1) when
  `MBEDTLS_USE_PSA_CRYPTO` is enabled - this was tasks like #5795, #5796,
  #5797. It is being done in PK and RSA PKCS#1 v1.5 by PR #6065.
- Step 3 was mostly not started at all before 3.2; it is being done for PK by
  PR #6065.

**Strategy for step 1:**

Regarding PK, X.509, and TLS, this is mostly achieved with only a few gaps.
(The strategy was outline in the previous section.)

Regarding libmbedcrypto, including the PSA Crypto core, this has not been
studied yet. For dependencies outside the PSA Crypto code (such as RSA
PKCS#1 v2.1 depending on MD), it should be checked whether this can be
achieved without backwards compatibility issues (currently applications can
call `mbedtls_rsa_xxx()` functions without calling `psa_crypto_init()` first),
otherwise a new compile-time option might be needed. For dependencies in the
PSA Crypto core, splitting `psa_crypto_init()` might be a topic (which might
also help for dependencies outside the core), with likely questions about
ordering (can we initialize drivers before the RNG or do some divers expect a

> working RNG?) and trying to avoid circular dependencies.

Author: And now I'm trying to image reasons why an actual scuba diver would need a RNG ;)

**Strategy for step 2:**

The most satisfying situation here is when we can just use the PSA Crypto API
for information management as well. However sometimes it may not be
convenient, for example in parts of the code that accept old-style identifier
(such as `mbedtls_md_type_t`) in their API and can't assume PSA to be
compiled in (such as `rsa.c`).

It is suggested that, as a temporary solution until we clean this up
later when removing the legacy API including its identifiers (G4), we may
occasionally use ad-hoc internal functions, such as the one introduced by PR
6065 in `library/md_internal.h`.

An alternative would be to have two different code paths depending on whether
`MBEDTLS_PSA_CRYPTO_C` is defined or not. However this is not great for
readability or testability.

**Strategy for step 3:**

There are currently two (competing) ways for crypto-using code to check if a
particular algorithm is supported: using `MBEDTLS_xxx` macros, and using
`PSA_WANT_xxx` macros. For example, PSA-based code that want to use SHA-256
will check for `PSA_WANT_ALG_SHA_256`, while legacy-based code that wants to
use SHA-256 will check for `MBEDTLS_SHA256_C` if using the `mbedtls_sha256`

> API, or for `MBEDTLS_MD_C && MBEDTLS_SHA256_C` if suing the `mbedtls_md` API.

Author: Indeed, there are things I dislike about this API but suing it is probably not gonna work.

- Cipher (G1): replace calls at each call site
- MD (G1): replace calls at each call site

Supporting builds with drivers without the software implementation
==================================================================

This section presents a plan towards G5: save code size by compiling out our
software implementation when a driver is available.

Additionally, we want to save code sive by compiling out the
abstractions layers that we are not using when `MBEDTLS_USE_PSA_CRYPTO` is
enabled (see previous section): MD and Cipher.

Let's expand a bit on the definition of the goal: in such a configuration
(driver used, software implementation and abstraction layer compiled out),
we want:

a. the library to build in a reasonably-complete configuration,
b. with all tests passing,

> c. and no more tests skipped than the same configuration with software

Reviewer: `No more` suggests that we're after numbers (so we can skip some, but add others), but I think that the idea here is to `have given features tested to the same extent using both configuration options`.

Reviewer: It's clarified in the next paragraph, so I'm OK with that.

   implementation.

Criterion (c) ensures not only test coverage, but that driver-based builds are
at feature parity with software-based builds.

We can roughly divide the work needed to get there in the following steps:

0. Have a working driver interface for the algorithms we want to replace.

> 1. Have users of these algorithms call to PSA, not the legacy API, for all

Reviewer: Is it possible to achieve this without compromising on test coverage? Do we do this 1:1 when switching to PSA API calls?
There are two reasons I'm asking:
- I can imagine some changes in data format that would result in tests being obsolete, or requiring more testing due to data format translation;
- It would be good to specify if we require the test coverage remaining the same after each step, or maybe after the whole work on a module is done. Or maybe we can just mention that there will be a "best effort" done to ensure that no substantial tests are omitted.

Author: I think work on a module can't be considered complete until test coverage is up to a satisfying level - for some reasonable value of satisfying.

For example, I consider the test coverage for PK in this PR to be satisfying, as the only tests skipped are those for pkparse that depend on PKCS5 and PKCS12, which represent a small fraction of the overall tests (less than 10%), and are arguably more integration tests than unit tests (handling this kind of encrypt keys is a feature of PEM, not PK).

As a counter-example, the test coverage for RSA in this PR is not satisfying IMO with 117 / 176 tests skipped in `test_suite_rsa` vs 0 skipped in the reference config. So, I do not consider the work on RSA complete.

Another counter-example that doesn't show in this PR: actually what I first tried was to make X.509 pass with drivers. This could be achieved by disabling RSA, leaving only EC certs to test with. Turned out this caused a very large proportion of tests to be skipped (I don't remember, 60% to 90% I think) and I considered this not enough, so I realised I had to work on RSA first, at least PKCS#1 v1.5.

We could test TLS 1.2 with only the PKCS#1 v1.5 part of RSA, because that's the only version it uses, but TLS 1.3 only uses PKCS#1 v2.1, so we couldn't consider it sufficiently tested if we didn't enable PKCS#1 v2.1 in the build... and so on.

(This is one of the reason I'm pushing an update with PKCS#1 v2.1 (incomplete as unit tests are missing so far), the other reason is this is an example of how I think we can deal with dependencies inside libmbecrypto (except for the PSA core's RNG) in a backwards-compatible way.)

Reviewer: I'm just afraid that we might lose this information and, since there's no step planned to analyse and make sure that the coverage is the same after all the work done (or, at least good enough) - any follow-up tasks might be lost. 
Do you think that it would be worth to add an additional analysis step after the work is done to determine if the coverage is satisfying? The document puts some pressure on `c. and no more tests skipped than the same configuration with software implementation.`, but the transition to PSA itself should also have such requirement.

**Strategy for step 3:**

There are currently two (complementary) ways for crypto-using code to check if a
particular algorithm is supported: using `MBEDTLS_xxx` macros, and using
`PSA_WANT_xxx` macros. For example, PSA-based code that want to use SHA-256
will check for `PSA_WANT_ALG_SHA_256`, while legacy-based code that wants to
use SHA-256 will check for `MBEDTLS_SHA256_C` if using the `mbedtls_sha256`
API, or for `MBEDTLS_MD_C && MBEDTLS_SHA256_C` if suing the `mbedtls_md` API.

It is suggested to introduce a new set of macros, `MBEDTLS_USE_PSA_WANT_xxx`,
for use in the parts of the code that use either API depending of whether
`MBEDTLS_USE_PSA_CRYPTO` is enabled (that is, PK, X.509 and TLS 1.2). This is
done for hash algorithms, as an example, by PR 6065. These macros can be used
in library code (most useful when algorithm availability is check far from the
code that will be using it, such as in TLS negotiation) as well as test
dependencies.

It should also be noted that there is a fourth case: utility functions /
information tables that are not tied to a particular crypto API, and may be
used by functions that are either purely PSA-based, purely legacy-based, or

> hybrid governed by `MBEDTL_USE_PSA_CRYPTO` should use `MBEDTLS_xxx ||

Reviewer: ```suggestion
hybrid governed by `MBEDTLS_USE_PSA_CRYPTO` should use `MBEDTLS_xxx ||
```

PSA_WANT_xxx` - for example, `oid_md_alg` from `oid.c`, used by both X.509 and

> RSA. A new family of macros `MBEDTLS_OR_PSA_WANT_xxx` is defined for this.

Reviewer: Things are starting to get a little bit complicated here:
- If `PSA_WANT_XXX` is defined, it usually turns on an `MBEDTLS_XXX` feature in `config_psa.h`;
- If `MBEDTLS_XXX` feature is defined, it usually adds a `PSA_WANT_XXX` define for a corresponding feature;
- If either of them will be defined, `MBEDTLS_OR_PSA_WANT_xxx` will be added for convenience;
If we go for the "partial" builds for information management only that you mentioned before, things get even more complicated, and these defines would need to get more fine-grained: `PSA_WANT_XXX_INFO`?.
Maybe it's just better to stay with `MBEDTLS_xxx || PSA_WANT_xxx` for clarity...

of the software implementation.

Note: the fact that step 1 is not achieved for all of libmbedcrypto (see
below) is the reason why criterion (a) has "a reasonably-complete
configuration", to allow working around internal crypto dependencies when
working on other parts such as X.509 and TLS - for example, a configuration
without RSA PKCS#1 v2.1 still allows reasonable use of X.509 and TLS.

Note: this is a conceptual division that will sometimes translate to how the
work is divided into PRs, sometimes not. For example, in situations where it's
not possible to achieve good test coverage at the end of step 1 or step 2, it
is preferable to group with the next step(s) in the same PR until good test
coverage can be reached.

**Status as of Mbed TLS 3.2:**

- Step 0 is achieved for most algorithms, with only a few gaps remaining.
- Step 1 is achieved for most of PK, X.509, and TLS when
  `MBEDTLS_USE_PSA_CRYPTO` is enabled with only a few gaps remaining (see
  docs/use-psa-crypto.md).
- Step 1 is not achieved for a lot of the crypto library including the PSA
  core. For example, `entropy.c` calls the legacy API

> `mbedtls_sha256` (or `mbedtls_sha512` optionally); `hmac_drbg.c` calls the

Reviewer: Indentation change

those configurations, and in configurations were the build was possible,
application code keeps working unchanged. An incomplete example of applying
this strategy, for RSA PKCS#1 v2.1, is here:
<https://github.com/mpg/mbedtls/pull/4>.

There is a problem with the modules used for the PSA RNG, as currently the RNG
is initialized before drivers and the key store. This part will need further
study, but in the meantime we can proceed with everything that's not the
entropy module of one of the DRBG modules, and that does not depend on one of
those modules.

**Strategy for step 2:**

The most satisfying situation here is when we can just use the PSA Crypto API
for information management as well. However sometimes it may not be
convenient, for example in parts of the code that accept old-style identifiers
(such as `mbedtls_md_type_t`) in their API and can't assume PSA to be
compiled in (such as `rsa.c`).

It is suggested that, as a temporary solution until we clean this up
later when removing the legacy API including its identifiers (G4), we may
occasionally use ad-hoc internal functions, such as the ones introduced by PR

> 6065 in `library/hash_info.[ch]`.

Reviewer: ```suggestion
#6065 in `library/hash_info.[ch]`.
```

Author: It's at the beginning of a line, so I'm afraid markdown will interpret it as a title. (My editor's syntax highlighting does.) That's why I went with "PR 6065" - I think "PR" makes it clear without the hash sign.

In order to cater to these new needs, new families of macros are introduced in
`library/legacy_or_psa.h`, see its documentation for details.

It should be noted that there are currently:
- too many different ways of computing a hash (low-level, MD, PSA);
- too many different ways to configure the library that influence which of
  these ways is available and will be used (`MBEDTLS_USE_PSA_CRYPTO`,
`MBEDTLS_PSA_CRYPTO_CONFIG`, `mbedtls_config.h` + `psa/crypto_config.h`).

As a result, we need more families of dependency macros than we'd like to.
This is a temporary situation until we move to a place where everthing is
based on PSA Crypto. In the meantime, long and explicit names where chosen for
the new macros in the hope of avoiding confusion.

Executing step 3 will mostly consist of using the right dependency macros in
the right places (once the previous steps are done).

**Note on testing**

Since supporting driver-only builds is not about adding features, but also
supporting existing features in new types of builds, the existing tests should
be enough to cover everything, as long as they're run in the newly supported

> configurations. An all.sh component needs to be present each main type of

Reviewer: ```suggestion
configurations. An all.sh component needs to be present in each main type of
```

> configutation we support (for example, with this (family of) algorithm

Reviewer: ```suggestion
configutations we support (for example, with this (family of) algorithms implemented using
```

driver-only, with or without `USE_PSA`).

There is however a risk, especially in step 3 where we change how dependencies
are expressed (sometimes in bulk), to get things wrong in a way that would
result in more tests being skipped, which is easy to miss. Care must be
taken to ensure this does not happen. The following criteria can be used:

- the sets of tests skipped in the default config and the full config must be
  the same before and after the PR that implements step 3;
- the set of tests skipped in the driver-only build is the same as in an
  equivalent software-based configuration, or the difference is small enough,

> justified, and a github issue is created to track it.

Reviewer: ```suggestion
  justified, and a github issue is created to track it.
```

An alternative would be to have two different code paths depending on whether
`MBEDTLS_PSA_CRYPTO_C` is defined or not. However this is not great for
readability or testability.

**Strategy for step 3:**

There are currently two (complementary) ways for crypto-using code to check if a
particular algorithm is supported: using `MBEDTLS_xxx` macros, and using
`PSA_WANT_xxx` macros. For example, PSA-based code that want to use SHA-256
will check for `PSA_WANT_ALG_SHA_256`, while legacy-based code that wants to
use SHA-256 will check for `MBEDTLS_SHA256_C` if using the `mbedtls_sha256`
API, or for `MBEDTLS_MD_C && MBEDTLS_SHA256_C` if using the `mbedtls_md` API.

Code that obeys `MBEDTLS_USE_PSA_CRYPTO` will want to use one of the two
dependencies above depending on whether `MBEDTLS_USE_PSA_CRYPTO` is defined:
if it is, the code want the algorithm available in PSA, otherwise, it wants it
available via the legacy API(s) is it using (MD and/or low-level).

The strategy for steps 1 and 2 above will introduce new situations: code that
currently compute hashes using MD (resp. a low-level hash module) will gain
the ability to "fall back" to using PSA if the legacy dependency isn't

> available. Data related to a certain hash (OID, sizes, translations) should

Reviewer 2: About OID specifically, we've been thinking of moving `oid.c` out of the crypto library, since it's mostly used by X.509. Crypto only uses OIDs for RSA PKCS#1v1.5 signatures and for pk parsing/writing, and that's not a lot of them. It's been an issue in the past that applications that only use RSA and not X.509 still need to include the whole `oid.o` instead of just one or two hash OIDs.

So I think a good follow-up would be to add the hash OIDs to the `hash_info` module.

only be included in the build if it is possible to use that hash in some way.

In order to cater to these new needs, new families of macros are introduced in
`library/legacy_or_psa.h`, see its documentation for details.

It should be noted that there are currently:

> - too many different ways of computing a hash (low-level, MD, PSA);

Reviewer 2: [PSA driver support is experimental, so no, that wouldn't be an API break](https://github.com/Mbed-TLS/mbedtls/issues/6469#issuecomment-1287830668).

> - too many different ways of computing a hash (low-level, MD, PSA);

Reviewer 2: The more I interact with the `hash_info` interface, the more I think it's just making the code base more complicated, and we should stick to `md.h` as the sole abstraction for code that isn't pure PSA.

the new macros in the hope of avoiding confusion.

Executing step 3 will mostly consist of using the right dependency macros in
the right places (once the previous steps are done).

**Note on testing**

Since supporting driver-only builds is not about adding features, but about
supporting existing features in new types of builds, testing will not involve
adding cases to the test suites, but instead adding new components in `all.sh`
that build and run tests in newly-supported configurations. For example, if
we're making some part of the library work with hashes provided only by
drivers when `MBEDTLS_USE_PSA_CRYPTO` is defined, there should be a place in
`all.sh` that builds and run tests in such a configuration.

There is however a risk, especially in step 3 where we change how dependencies
are expressed (sometimes in bulk), to get things wrong in a way that would
result in more tests being skipped, which is easy to miss. Care must be
taken to ensure this does not happen. The following criteria can be used:

- the sets of tests skipped in the default config and the full config must be
  the same before and after the PR that implements step 3;

> - the set of tests skipped in the driver-only build is the same as in an

Reviewer: I think that assuming the reviews are ideal, no such issue would happen. But sometimes, when renaming a lot of dependencies `en masse`, mistakes hard to spot might happen. Example:
A `.data` file test dependency on `MBEDTLS_SHA256_C` is bulk changed to `MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA`, since we aim to provide the same functionality via either of these configuration sets. Such rename is applied to 300 tests, and an initial check by the commiter doesn't show anything suspicious. He/she doesn't check the number of tests skipped.
As it turns out, one of the tests had a following dependency: `!MBEDTLS_SHA256_C` which was changed to `!MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA`. What if we didn't have the configuration combination with no MD and without PSA? Such test would be always skipped.
Maybe this exact example isn't accurate, since we have a build that would cover this case, but maybe there are some that we didn't check.
The idea behind this "skipped tests" script is to double check that there were no missed cases, regardless of the reason.

> - the set of tests skipped in the driver-only build is the same as in an

Reviewer 2: Good point. Or: do a bulk renaming, then a merge/rebase brings in some new test cases for which the renaming wasn't applied.

Still, I don't understand why we'd want to do this comparison, as opposed to checking that there are no longer any old-style dependencies in places that should only have new-style dependencies.

> - the set of tests skipped in the driver-only build is the same as in an

Reviewer 3: For example while working on `Driver-only hashes: X.509` I changed dependencies in `tests/suites/test_suite_x509write.data`. Replaced `MBEDTLS_MD5_C` and `MBEDTLS_SHAxxx_C` with the corresponding `MBEDTLS_HAS_ALG_xxx_VIA_MD_OR_PSA` macros.

Then when I run `component_test_psa_crypto_config_accel_hash_use_psa` from `all.sh` it turned out that set of `> Certificate write check Server1 SHA1` tests were skipped.
This was because there was also `MBEDTLS_SHA1_C` dependency in `.function` file.

It is easy to miss something. I found `outcome-analysis.sh` script very useful.

Executing step 3 will mostly consist of using the right dependency macros in
the right places (once the previous steps are done).

**Note on testing**

Since supporting driver-only builds is not about adding features, but also
supporting existing features in new types of builds, the existing tests should
be enough to cover everything, as long as they're run in the newly supported
configurations. An all.sh component needs to be present each main type of

> configutation we support (for example, with this (family of) algorithm

Reviewer: This sentence is a bit hard to understand. Is an `all.sh` component an entity that we use anywhere outside than in `all.sh`?  What does it mean that it has to be present in each main types of configurations? 
Example: `component_test_full_cmake_gcc_asan` - does this (in reference to this sentence) mean that it should work with both the drivers and software implementations?

Author: Ah, sorry, that was unclear indeed. I'm not trying to say that each `all.sh` component should work with drivers, I'm trying to say that when we add support for X provided by drivers, we need to test that somewhere in `all.sh`. So, basically, that this PR had to add `component_test_psa_crypto_config_accel_hash_use_psa` - the goal of the PR is to make some things work with driver-only hashes and `USE_PSA`, so it has to test a build with driver-only hashes and `USE_PSA`.

So really this paragraph is not saying more than "Things should be tested. New configurations are not tested by adding test cases in the test suites, but by adding components in `all.sh`."
